File: jpa-fetch-join/src/main/java/com/example/jpafetchjoin/order/OrderRepository.java
package com.example.jpafetchjoin.order;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;

public interface OrderRepository extends JpaRepository<Order, Long> {


  @Query(
      value = "select  o from Order o "
          + "left join fetch o.addresses "
          + "left join fetch o.products ",
      countQuery = "select count(o) from Order o"
  )
  Page<Order> findByPageWithAll(Pageable pageable);


  @Query(
      value = "select  o from Order o "
          + "left join fetch o.addresses ",
      countQuery = "select count(o) from Order o"
  )
  Page<Order> findByPageWithAddress(Pageable pageable);


  @Query(
      value = "select  o from Order o "
          + "left join fetch o.products ",
      countQuery = "select count(o) from Order o"
  )
  Page<Order> findByPageWithProducts(Pageable pageable);

}

File: jpa-n-plus-1/src/main/kotlin/com/example/jpanplus1/order/OrderRepository.kt
package com.example.jpanplus1.order

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

interface OrderRepository : JpaRepository<Order, Long> {

    @Query("select o from Order o join fetch o.member")
    fun findAllOrder(): List<Order>

    @Query("select o from Order o where o.id = :id")
    fun findByIdCustom(@Param("id") id: Long): Order

    fun findByNumber(number: String): Order
}